homeassistant/components/media_player/cast.py

```diff
@@ -33,90 +38,159 @@
     SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_PREVIOUS_TRACK | \
     SUPPORT_NEXT_TRACK | SUPPORT_PLAY_MEDIA | SUPPORT_STOP | SUPPORT_PLAY
 
-KNOWN_HOSTS_KEY = 'cast_known_hosts'
+INTERNAL_DISCOVERY_RUNNING = 'cast_discovery_running'
+# UUID -> CastDevice mapping; None key has all CastDevices without UUID
+ADDED_CAST_DEVICES_KEY = 'cast_added_cast_devices'
+# Stores every discovered pychromecast.Chromecast
+KNOWN_CHROMECASTS_KEY = 'cast_all_chromecasts'
+
+SIGNAL_CAST_DISCOVERED = 'cast_discovered'
 
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
     vol.Optional(CONF_HOST): cv.string,
     vol.Optional(CONF_IGNORE_CEC): [cv.string],
 })
 
 
-# pylint: disable=unused-argument
-def setup_platform(hass, config, add_devices, discovery_info=None):
+def _setup_internal_discovery(hass: HomeAssistantType) -> None:
+    """Set up the pychromecast internal discovery."""
+    if hass.data.get(INTERNAL_DISCOVERY_RUNNING, False):
+        # Internal discovery is already running
+        return
+
+    import pychromecast
+
+    def internal_callback(name):
+        """Called when zeroconf has discovered a new chromecast."""
+        try:
+            chromecast = pychromecast._get_chromecast_from_host(
+                listener.services[name], blocking=True)
```

Comment:
Using a protected function here as a temporary hack. We need some way of using the *non-blocking* discovery with *blocking* Chromecast objects. If we want to merge this PR, I would of course create a PR in pychromecast to allow this without a "hack".

Reply:
Additionally, there's another potential problem here (that was also there before). Every time we find a chromecast, we create a full `Chromecast` object, with a separate `Thread` object, `socket` and whatnot. (related: #11945)

We might be able to get away in some cases by just matching the host, port and information we get from zeroconf.